### ISSUE TITLE
Kubelet to POST pod status to apiserver

### ISSUE THREAD

**Issue by erictune** [GoogleCloudPlatform/kubernetes#4561]
Split off #156 as a smaller more specific work item.

In the kubelet, once every N sync loops, it should POST `/api/$VERSION/namespaces/$NS/pods/$NAME/status` for each pod. 

The kubelet would do this if enabled by a flag, and emit a warning if it failed to POST the update.

The kubelet would ideally handle a 429 by retrying after the `Retry-after` header.

**Comment by dchen1107** [comment]
In a long run, we want to do a bulk POST to apiserver, but that is not v1 blocker.

**Comment by derekwaynecarr** [comment]
api/$VERSION/namespaces/$NS/pods/$NAME/status for each pod.

**Comment by fgrzadkowski** [comment]
As suggested by @dchen1107 I'll work on this. 

<For some reason I can't assign myself to this issue>

**Comment by timothysc** [comment]
Out of curiosity why can we roll up Node and Pod status into a single status update?

**Comment by dchen1107** [comment]
That is the plan to have a bulk status update, but not strictly required at this moment.

**Comment by erictune** [comment]
@timothysc what URL do you propose for doing a node + pod update?

**Comment by yujuhong** [comment]
@fgrzadkowski, FYI, my PR #5019 modifies kubelet to reject (set pod status to fail) pods that have port conflict. The status is stored in a map and gets reported back later via status polling.

**Comment by smarterclayton** [comment]
#5085 is quasi blocked on this - in that if we do graceful deletion with TTL (the optimal way) then we won't be able to clear the binding at the point the pod is actually deleted.  We could still delete the binding at the point the TTL starts, which is somewhat reasonable (since you can't stop or delay a deletion as I've implemented it so far) because it will trigger the kubelet to remove the pod gracefully.  However, since true graceful would be SIGTERM to Docker with the remaining TTL window as soon as the pod sees it, then SIGKILL when delete happens, that's harder to do if the pod disappears from the binding.

**Comment by fgrzadkowski** [comment]
I have almost ready PR for this (some tests are failing). Will send it on Monday.

**Comment by fgrzadkowski** [comment]
I had to revert PR #5305 due to bugs. Reopening issue. Will send fixed version soon.

**Comment by timothysc** [comment]
So in the case of deletion, we're now seeing are is gobs traffic on deletion trying to send updated status and the api server presenting NOT-FOUND. 

Easy repro: 
1. run density tests
2. start traffic monitoring on api-server  ** tcpdump -nnvvXSs 1514 'tcp port 8080 and (((ip[2:2] - ((ip[0]&0xf)<<2)) - ((tcp[12]&0xf0)>>2)) != 0)' ** 
3. Wait for cleanup..

**Comment by vmarmol** [comment]
Sent out #5619 to lower and spread that load. It lowers the qps from 100 to ~9.

**Comment by vmarmol** [comment]
After discussions with @bgrant0607 and @dchen1107, they suggested to only update the status when it changes and on startup. The heartbeat will be handled by the node controller rather than per-pod. I'll file a separate issue for that and #5619  will go in for now.
